### PR TITLE
[Feature] Pools table bookmarks

### LINKED ISSUE #10269: ✨ Bookmark processes for recruiters
## User story
As a pool operator (recruiter), I want a quick way to find processes I work with frequently, so I don't have to search through the processes table everytime.

## Details
- We are proposing adding a "bookmark" feature to the processes table. This would sticky at the top of the table (page 1) the processes that they had bookmarked.
- These bookmarks would be (PER USER) meaning It would only be for that user. Not for all users like we do with candidates in a process.

## Screenshot
[Figma file](https://www.figma.com/file/j7fxenmvTCdM7T5aoQCvVG/Admin---Recruitment-index?type=design&node-id=3%3A19056&mode=design&t=1d1TkV7ko9xkw3Wn-1), designer @Jerryescandon 

Image

## Acceptance criteria
- [ ] Processes can be bookmarked on the process table
- [ ] Bookmarks are saved by user (and not for all users)
- [ ] Bookmarked processes always show up at the top of page 1 of the process table
- [ ] Create pivot table btw users and pools

### PR DESCRIPTION
**Pull request by yonikid15**
🤖 Resolves #10269

## 👋 Introduction

**This is a draft until the ordering works correctly.**

- Adds bookmarking feature to pools table.
- Adds a pivot table btw pool and users (bookmarks).
- The bookmarks are unique to each user.
- The bookmarks should stick to top of table (still working on this).

## 🕵️ Details

Add any additional details that could assist with reviewing or testing this PR.

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. Build app
2. Login as admin
3. Go to view processes on admin dashboard
4. Ensure bookmarking feature works along with other table tools.

### CODE REVIEW

**Comment by vd1992** [comment]
Where is sorting failing? It seems to work for me. 
Doesn't appear to do a secondary sort though, sorting a second way after bookmark, so sort by bookmark and then say stream

File: api/graphql/schema.graphql [hunk 1]
```diff
@@ -1953,6 +1963,13 @@ type Mutation {
     @update
     @guard
     @canFind(ability: "updatePublished", find: "id")
+  togglePoolUserBookmark(
+    userId: UUID! @rename(attribute: "user_id")
+    poolId: UUID! @rename(attribute: "pool_id")
+  ): Pool
+    @guard
+    @canFind(ability: "view", find: "pool_id", model: "Pool", injectArgs: true)
```
Comment (vd1992) [inline api/graphql/schema.graphql:1971]: This allows anyone to update anyone else's bookmark which we probably don't want. I think it is reasonable for someone to only be able to update their own bookmarks.

I think that can be easily done like so below using `inject`, where the acting/auth context object feeds into the `args.user_id` value

image

**Review by vd1992 — COMMENTED** [review]
It seems to work well, I can't comment on if the sorting is ideal since I didn't get what was wrong. Did find something important below

File: api/graphql/schema.graphql [hunk 2]
```diff
@@ -1962,6 +1972,11 @@ type Mutation {
     @update
     @guard
     @canFind(ability: "updatePublished", find: "id")
+  togglePoolUserBookmark(poolId: UUID! @rename(attribute: "pool_id")): Pool
+    @guard
+    @inject(context: "user.id", name: "user_id")
```
Comment (vd1992) [inline api/graphql/schema.graphql:1977]: You are injecting `user_id` here

File: api/app/GraphQL/Mutations/TogglePoolUserBookmark.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+     * Toggles a user's bookmarked pool
+     *
+     * @param  array{}  $args
+     */
+    public function __invoke($_, array $args)
+    {
+        /** @var \App\Models\User */
+        $user = Auth::user();
+        $pool = Pool::find($args['pool_id']);
+        $user->poolBookmarks()->toggle($pool->id);
```
Comment (vd1992) [inline api/app/GraphQL/Mutations/TogglePoolUserBookmark.php:21]: But do not use `user_id` anywhere

I don't know what the safe/best way would be though. The highlighted above or finding from the injected `user_id`
Any wisdom to share on this @tristan-orourke  ?

File: api/graphql/schema.graphql [hunk 4]
```diff
@@ -815,10 +823,12 @@ type Query {
     includeIds: [UUID!] @in(key: "id")
     excludeIds: [UUID!] @notIn(key: "id")
     where: PoolFilterInput
+    orderByPoolBookmarks: PoolBookmarksOrderByInput @scope
     orderByTeamDisplayName: PoolTeamDisplayNameOrderByInput @scope
     orderBy: _
       @orderBy(
         relations: [
+          # { relation: "poolBookmarks", columns: ["created_at"] }
```
Comment (vd1992) [inline api/graphql/schema.graphql:831]: Commented out

File: apps/web/src/pages/Pools/IndexPoolPage/components/helpers.tsx [hunk 5]
```diff
@@ -190,14 +195,27 @@ export function getOrderByClause(
...
 
+  // const poolBookmarksOrderByClause: QueryPoolsPaginatedOrderByRelationOrderByClause =
+  //   {
+  //     column: undefined,
+  //     order: SortOrder.Asc,
+  //     poolBookmarks: {
+  //       aggregate: OrderByRelationWithColumnAggregateFunction.Max,
+  //       column: "CREATED_AT" as QueryPoolsPaginatedOrderByPoolBookmarksColumn,
+  //     },
+  //   };
```
Comment (vd1992) [inline apps/web/src/pages/Pools/IndexPoolPage/components/helpers.tsx:210]: Commented out

**Review by vd1992 — COMMENTED** [review]
Got some commented out bits below, did you mean to keep them around?
I'd also like to let someone else handle final approval to judge sorting

**Comment by yonikid15** [comment]
Sorry Vachan I might have accidentally clicked the re-review.  However, I just addressed the comments. For the other comment, I decided to get the user form Auth, however we can see what Tristans comment is on that.

File: apps/web/src/pages/Pools/IndexPoolPage/components/helpers.tsx (hunk 5 above)
Comment (yonikid15) [reply apps/web/src/pages/Pools/IndexPoolPage/components/helpers.tsx]: Fixed by [79cc264](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10509/commits/79cc264065ed10aee21f6ca7574f7031f338df8e)

File: api/graphql/schema.graphql (hunk 2 above)
Comment (yonikid15) [reply api/graphql/schema.graphql]: See [79cc264](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10509/commits/79cc264065ed10aee21f6ca7574f7031f338df8e)

File: api/app/GraphQL/Mutations/TogglePoolUserBookmark.php (hunk 3 above)
Comment (yonikid15) [reply api/app/GraphQL/Mutations/TogglePoolUserBookmark.php:21]: See [79cc264](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10509/commits/79cc264065ed10aee21f6ca7574f7031f338df8e)

Comment (yonikid15) [reply api/app/GraphQL/Mutations/TogglePoolUserBookmark.php:21]: I talked with Tristan and he said this option looks good :+1:

File: api/graphql/schema.graphql [hunk 6]
```diff
@@ -815,6 +823,7 @@ type Query {
     includeIds: [UUID!] @in(key: "id")
     excludeIds: [UUID!] @notIn(key: "id")
     where: PoolFilterInput
+    orderByPoolBookmarks: PoolBookmarksOrderByInput @scope
```
Comment (petertgiles) [inline api/graphql/schema.graphql:826]: Why is a second query fired when the orderby is already satisfied in the first query?
image

**Review by petertgiles — CHANGES_REQUESTED** [review]
Very cool!  :sunglasses: This works very well already.  I left a few small ideas to level it up.  Let me know if you want to chat about any of them.

File: api/app/Models/User.php [hunk 7]
```diff
@@ -201,6 +202,11 @@ public function pools(): HasMany
         return $this->hasMany(Pool::class);
     }
 
+    public function poolBookmarks(): BelongsToMany
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply api/app/Models/User.php:205]: Add seeder in [16e44b9](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10509/commits/16e44b9a1d57925a0505ccdac1f6431891fff0ce)

File: api/app/GraphQL/Mutations/TogglePoolUserBookmark.php [hunk 8]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+namespace App\GraphQL\Mutations;
+
+use App\Models\Pool;
+use Illuminate\Support\Facades\Auth;
+
+final class TogglePoolUserBookmark
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply api/app/GraphQL/Mutations/TogglePoolUserBookmark.php:8]: Add test in [4b0b06b](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10509/commits/4b0b06b760ba4c8e024369f7d7ed701488ec6cf9)

File: api/graphql/schema.graphql (hunk 6 above)
Comment (yonikid15) [reply api/graphql/schema.graphql:827]: I wasn't seeing this, but let me double check.

File: api/app/Models/Pool.php [hunk 9]
```diff
@@ -468,6 +473,28 @@ public function scopeOrderByTeamDisplayName(Builder $query, ?array $args): Build
...
+    {
+        extract($args);
+
+        /** @var \App\Models\User */
+        $user = Auth::user();
+
+        // order the pools so that the bookmarks connected to current user sticks to the top
+        if ($order && $user) {
+            $query->orderBy(
+                $user->select('pool_user_bookmarks.created_at')
```
[the remark this answers is not in the document]
Comment (yonikid15) [reply api/app/Models/Pool.php:497]: I might need some more guidance on this. I removed take(1) and it works fine without it :+1: 
However, without ordering by the bookmarks `created_at` date, the order would vary causing the tests to fail.

[the remark this answers is not in the document]
Comment (petertgiles) [reply api/app/Models/Pool.php:497]: This is working OK for me, I think.  The order by created date in the orderBy parameter is keeping it steady.
image
We can look at this together if I'm missing something.

File: api/graphql/schema.graphql (hunk 6 above)
Comment (petertgiles) [reply api/graphql/schema.graphql:827]: I'm still seeing it but if it's not obvious to you I think we can let it go for later optimization.

File: api/app/Models/Pool.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (yonikid15) [reply api/app/Models/Pool.php]: Awesome thanks this solution is working :+1:

**Review by petertgiles — APPROVED** [review]
Awesome!  :1st_place_medal:

**Pull request merged**